src/rov_planning/lib/mosquitto_install/share/man/man7/mqtt.7
'\" t
.\"     Title: mqtt
.\"    Author: [see the "Author" section]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 08/02/2024
.\"    Manual: Conventions and miscellaneous
.\"    Source: Mosquitto Project
.\"  Language: English
.\"
.TH "MQTT" "7" "08/02/2024" "Mosquitto Project" "Conventions and miscellaneous"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
mqtt \- MQ Telemetry Transport
.SH "SYNOPSIS"
.HP \w'\fBMQTT\fR\ 'u
\fBMQTT\fR
.SH "DESCRIPTION"
.PP
\fBMQTT\fR
is a lightweight publish/subscribe messaging protocol\&. It is useful for use with low power sensors, but is applicable to many scenarios\&.
.PP
This manual describes some of the features of MQTT version 3\&.1\&.1/3\&.1, to assist end users in getting the most out of the protocol\&. For more complete information on MQTT, see
http://mqtt\&.org/\&.
.SH "PUBLISH/SUBSCRIBE"
.PP
The MQTT protocol is based on the principle of publishing messages and subscribing to topics, or "pub/sub"\&. Multiple clients connect to a broker and subscribe to topics that they are interested in\&. Clients also connect to the broker and publish messages to topics\&. Many clients may subscribe to the same topics and do with the information as they please\&. The broker and MQTT act as a simple, common interface for everything to connect to\&. This means that you if you have clients that dump subscribed messages to a database, to Twitter, Cosm or even a simple text file, then it becomes very simple to add new sensors or other data input to a database, Twitter or so on\&.
.SH "TOPICS/SUBSCRIPTIONS"
.PP
Messages in MQTT are published on topics\&. There is no need to configure a topic, publishing on it is enough\&. Topics are treated as a hierarchy, using a slash (/) as a separator\&. This allows sensible arrangement of common themes to be created, much in the same way as a filesystem\&. For example, multiple computers may all publish their hard drive temperature information on the following topic, with their own computer and hard drive name being replaced as appropriate:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
sensors/COMPUTER_NAME/temperature/HARDDRIVE_NAME
.RE
.PP
Clients can receive messages by creating subscriptions\&. A subscription may be to an explicit topic, in which case only messages to that topic will be received, or it may include wildcards\&. Two wildcards are available,
\fB+\fR
or
\fB#\fR\&.
.PP
\fB+\fR
can be used as a wildcard for a single level of hierarchy\&. It could be used with the topic above to get information on all computers and hard drives as follows:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
sensors/+/temperature/+
.RE
.PP
As another example, for a topic of "a/b/c/d", the following example subscriptions will match:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
a/b/c/d
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
+/b/c/d
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
a/+/c/d
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
a/+/+/d
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
+/+/+/+
.RE
.PP
The following subscriptions will not match:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
a/b/c
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
b/+/c/d
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
+/+/+
.RE
.PP
\fB#\fR
can be used as a wildcard for all remaining levels of hierarchy\&. This means that it must be the final character in a subscription\&. With a topic of "a/b/c/d", the following example subscriptions will match:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
a/b/c/d
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
#
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
a/#
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
a/b/#
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
a/b/c/#
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
+/b/c/#
.RE
.PP
Zero length topic levels are valid, which can lead to some slightly non\-obvious behaviour\&. For example, a topic of "a//topic" would correctly match against a subscription of "a/+/topic"\&. Likewise, zero length topic levels can exist at both the beginning and the end of a topic string, so "/a/topic" would match against a subscription of "+/a/topic", "#" or "/#", and a topic "a/topic/" would match against a subscription of "a/topic/+" or "a/topic/#"\&.
.SH "QUALITY OF SERVICE"
.PP
MQTT defines three levels of Quality of Service (QoS)\&. The QoS defines how hard the broker/client will try to ensure that a message is received\&. Messages may be sent at any QoS level, and clients may attempt to subscribe to topics at any QoS level\&. This means that the client chooses the maximum QoS it will receive\&. For example, if a message is published at QoS 2 and a client is subscribed with QoS 0, the message will be delivered to that client with QoS 0\&. If a second client is also subscribed to the same topic, but with QoS 2, then it will receive the same message but with QoS 2\&. For a second example, if a client is subscribed with QoS 2 and a message is published on QoS 0, the client will receive it on QoS 0\&.
.PP
Higher levels of QoS are more reliable, but involve higher latency and have higher bandwidth requirements\&.
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
0: The broker/client will deliver the message once, with no confirmation\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
1: The broker/client will deliver the message at least once, with confirmation required\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
2: The broker/client will deliver the message exactly once by using a four step handshake\&.
.RE
.SH "RETAINED MESSAGES"
.PP
All messages may be set to be retained\&. This means that the broker will keep the message even after sending it to all current subscribers\&. If a new subscription is made that matches the topic of the retained message, then the message will be sent to the client\&. This is useful as a "last known good" mechanism\&. If a topic is only updated infrequently, then without a retained message, a newly subscribed client may have to wait a long time to receive an update\&. With a retained message, the client will receive an instant update\&.
.SH "CLEAN SESSION / DURABLE CONNECTIONS"
.PP
On connection, a client sets the "clean session" flag, which is sometimes also known as the "clean start" flag\&. If clean session is set to false, then the connection is treated as durable\&. This means that when the client disconnects, any subscriptions it has will remain and any subsequent QoS 1 or 2 messages will be stored until it connects again in the future\&. If clean session is true, then all subscriptions will be removed for the client when it disconnects\&.
.SH "WILLS"
.PP
When a client connects to a broker, it may inform the broker that it has a will\&. This is a message that it wishes the broker to send when the client disconnects unexpectedly\&. The will message has a topic, QoS and retain status just the same as any other message\&.
.SH "SEE ALSO"
\fBmosquitto\fR(8), \fBmosquitto_pub\fR(1), \fBmosquitto_sub\fR(1)
.SH "AUTHOR"
.PP
Roger Light
<roger@atchoo\&.org>


src/rov_planning/src/mqtt/src/mqtt.cpp
/**
 * @file mqtt.cpp
 * @brief
 * @author zeng panapn (zengpanpan@iat-auto.cn)
 * @version 1.0
 * @date 2022-04-07
 *
 * @copyright Copyright (c) 2022  阿尔特汽车技术股份有限公司
 *
 * Unpublished copyright. All rights reserved. This material contains
 * proprietary information that should be used or copied only within
 * IAT, except with written permission of IAT.
 *
 * @par 修改日志:
 * <table>
 * <tr><th>Date       <th>Version <th>Author  <th>Description
 * <tr><td>2022-04-07 <td>1.0     <td>Zeng Panpan     <td>Initialize create
 * </table>
 */

#include "mqtt.h"
namespace MQTT {

Mqtt_imp::Mqtt_imp(const char *id) 
    : mosquittopp(id), 
      closed_(true),
      connected_(false),
      pad_connected_(false) {}

Mqtt_imp &Mqtt_imp::get_single() {
    static Mqtt_imp instance("planning");
    return instance;
}

Mqtt_imp::~Mqtt_imp() { stop(); }

bool Mqtt_imp::init() {
    // 设置MQTT遗言消息
    will_set(WILL_MSG);
    // 设置订阅消息的主题
    string_topic_map_["SPAD/ModeControl"] = Topic::SPAD_MODE_CONTROL;
    string_topic_map_["SPAD/RCControl"] = Topic::SPAD_RC_CONTROL;
    string_topic_map_["SPAD/ArmedControl"] = Topic::SPAD_ARMED_CONTROL;
    string_topic_map_["SPAD/Heartbeat"] = Topic::SPAD_HEARTBEAT;
  	string_topic_map_["SPAD/AutoCruise"] = Topic::SPAD_AUTO_CRUISE;
  	string_topic_map_["SPAD/DebugTargetValue"] = Topic::SPAD_DEBUG_TARGET_VALUE;
	
    return true;
}

void Mqtt_imp::start() {
    closed_ = false;
    mqtt_thread_ = std::thread(&Mqtt_imp::process, this);
}

void Mqtt_imp::stop() {
    closed_ = true;
    try {
        if (mqtt_thread_.joinable()) {
            mqtt_thread_.detach();
        }
       
    } catch (const std::exception &e) {
        std::cout << "catched mqtt thread exception (shutdown):" << e.what() << std::endl;
    }
} 

void Mqtt_imp::on_connect(int rc) 
{ 
    if (rc == 0) {
      std::cout << "Broker connected! " << std::endl;
      connected_.store(true);
      subscribe(nullptr, "SPAD/RCControl");
      subscribe(nullptr, "SPAD/ModeControl");
      subscribe(nullptr, "SPAD/ArmedControl");
      subscribe(nullptr, "SPAD/Heartbeat");
	  subscribe(nullptr, "SPAD/AutoCruise");
	  subscribe(nullptr, "SPAD/DebugTargetValue");
    } else {
        std::cerr << "Connect failed: " << mosqpp::connack_string(rc) << std::endl;
    }
    
}

void Mqtt_imp::on_disconnect(int rc) 
{ 
    std::cout << "Disconnected: " << rc << ". Reconnecting..." << std::endl;
    connected_.store(false);
}

void Mqtt_imp::on_publish(int rc) {  }

void Mqtt_imp::on_subscribe(int mid, int qos_count, const int *granted_qos) {
    std::cout << "sub mid:" << mid << std::endl;
}

void Mqtt_imp::SetRCControlCallBack(const std::function<void(const RCControl&)>& cb)
{
    rc_control_func_ = cb;
}

void Mqtt_imp::SetModeControlCallBack(const std::function<void(const ModeControl&)>& cb)
{
    mode_control_func_ = cb;
}

void Mqtt_imp::SetArmedControlCallBack(const std::function<void(const ArmedControl&)>& cb)
{
    armed_control_func_ = cb;
}

void Mqtt_imp::SetDebugTargetValueCallBack(const std::function<void(int, double)>& cb)
{
    debug_target_value_func_ = cb;
}

void Mqtt_imp::SetMotionPlannerCallBack(const std::function<void(const RCControl&)>& cb)
{
    motion_planner_func_ = cb;
}

void Mqtt_imp::on_message(const struct mosquitto_message *message) {
    bool res = false;
    Topic topic = Topic::NO_TOPIC;
    for (const auto &temp : string_topic_map_) {
        mosqpp::topic_matches_sub(temp.first.c_str(), message->topic, &res);
        if (res) {
            topic = temp.second;
            break;
        }
    }
	 std::cout << (const char *)message->topic << std::endl;
     std::cout << (const char *)message->payload << std::endl;
    switch (topic) {
        case Topic::SPAD_RC_CONTROL: {
            ParseRCControl(static_cast<const char *>(message->payload));
            break;
        }
        case Topic::SPAD_MODE_CONTROL: {
            ParseModeControl(static_cast<const char *>(message->payload));
            break;
        }
        case Topic::SPAD_ARMED_CONTROL: {
            ParseArmedControl(static_cast<const char *>(message->payload));
            break;
        }
        case Topic::SPAD_HEARTBEAT: {            
            ParseHeartbeat(static_cast<const char *>(message->payload));
            break;
        }
		case Topic::SPAD_AUTO_CRUISE: {            
            ParseAutoCruise(static_cast<const char *>(message->payload));
            break;
        }
		case Topic::SPAD_DEBUG_TARGET_VALUE: {            
            DebugTargetValue(static_cast<const char *>(message->payload));
            break;
        }
        default:
            break;
    }
}


void Mqtt_imp::process() {
  
	const int delay_time{2};
    reconnect_delay_set(delay_time, delay_time, true);
    connect_async(SERVER_IP,PORT,  KEEP_ALIVE); // 异步连接
        
    // 循环处理数据
    loop_start();
    auto period = std::chrono::milliseconds(HEART_BEAT_ADD_PERIOD);
    while (!closed_) {
        {
            std::lock_guard<std::mutex> lock(heartbeat_mutex_);
            heartbeat_reset_count_++;
            if (heartbeat_reset_count_ >= MAX_HEARTBEAT_COUNT) {
                pad_connected_.store(false);
            }
        }
        
        /* Blocks the execution of the current thread until specified point of time (next_run) has been reached */
        std::this_thread::sleep_for(period);
        
    }
}


bool Mqtt_imp::ParseRCControl(const char *mess) {

    rapidjson::Document dom;
    RCControl temp;
    if (!dom.Parse(mess).HasParseError()) {
        if (dom.HasMember("rc_in") && dom["rc_in"].IsInt()) {
            temp.rc_in = dom["rc_in"].GetInt();
        }
        if (dom.HasMember("rc_control_value") && dom["rc_control_value"].IsDouble()) {
            temp.rc_control_value = dom["rc_control_value"].GetDouble();
        }
		auto_cruise_->CloseAutoCruiseThread();
		motion_planner_func_(temp);
        rc_control_func_(temp);
		
      
    } else {
        return false;
    }
    return true;
}

bool Mqtt_imp::ParseModeControl(const char *mess) {
    rapidjson::Document dom;
    ModeControl temp;
    if (!dom.Parse(mess).HasParseError()) {
        if (dom.HasMember("mode_control") && dom["mode_control"].IsInt()) {
            temp.mode_control = dom["mode_control"].GetInt();
        }
		auto_cruise_->CloseAutoCruiseThread();
        mode_control_func_(temp);
    } else {
        return false;
    }
    return true;

}

bool Mqtt_imp::ParseArmedControl(const char *mess) {
    rapidjson::Document dom;
    ArmedControl temp;
    if (!dom.Parse(mess).HasParseError()) {
        if (dom.HasMember("armed_control") && dom["armed_control"].IsInt()) {
            temp.armed_control = dom["armed_control"].GetInt();
        }
		auto_cruise_->CloseAutoCruiseThread();
        armed_control_func_(temp);
        
      
    } else {
        return false;
    }
    return true;

}

bool Mqtt_imp::ParseHeartbeat(const char *mess) {
    ROS_INFO("Received heartbeat from HMI.");
    {
        std::lock_guard<std::mutex> lock(heartbeat_mutex_);
        heartbeat_reset_count_ = 0U;
        pad_connected_.store(true);
    }

    return true;
}

bool Mqtt_imp::ParseAutoCruise(const char *mess) {
    ROS_INFO("Received auto cruise");
    auto_cruise_->StartAutoCruiseThread();
    return true;
}

bool Mqtt_imp::DebugTargetValue(const char *mess) {
    ROS_INFO("Received DebugTargetValue");
	rapidjson::Document dom;
	int type{0};
	double value{0.0};
    if (!dom.Parse(mess).HasParseError()) {
        if (dom.HasMember("type") && dom["type"].IsInt()) {
            type = dom["type"].GetInt();
        }
		if (dom.HasMember("value") && dom["value"].IsDouble()) {
            value = dom["value"].GetDouble();
        }
		auto_cruise_->CloseAutoCruiseThread();       
        debug_target_value_func_(type, value);
    } else {
        return false;
    }
    return true;
}

void Mqtt_imp::SetImuIfo(const ImuInfo& imu_info)
{
    std::lock_guard<std::mutex> lock(data_mutex_);
    imu_info_ = imu_info;
}

void Mqtt_imp::SetState(const ModeStatus& status) {
    
    rapidjson::StringBuffer buf_json;
    rapidjson::PrettyWriter<rapidjson::StringBuffer> writer(buf_json);
    {
        std::lock_guard<std::mutex> lock(data_mutex_);
        mode_status_ = status;
        writer.StartObject();
        writer.Key("motion_status");
        writer.Int(mode_status_.motion_status);
        writer.Key("armed_status");
        writer.Int(mode_status_.armed_status);
        writer.Key("depth");
        writer.Double(mode_status_.depth);
        writer.EndObject();

    }
    
    std::string result = buf_json.GetString();
    const char* topic = "MACH/ModeStatus";
    // ROS_INFO("motion:%d, armed:%d, depth:%f",status.motion_status, status.armed_status, status.depth);
    if (!connected_.load()) {
        return;
    }
    publish(nullptr, topic, result.size(), result.data());
}

void Mqtt_imp::SetRCOut(const std::vector<int>& data) {
   
    rapidjson::StringBuffer buf_json;
    rapidjson::PrettyWriter<rapidjson::StringBuffer> writer(buf_json);
    writer.StartObject();
    writer.Key("motion_control");
    writer.StartArray();
    // ROS_INFO("-------------------------------");
    for (const auto& value : data) {
        writer.Int(value);         
        // ROS_INFO("value:%d",value);
    }
    writer.EndArray();
    {
        std::lock_guard<std::mutex> lock(data_mutex_);
        writer.Key("yaw");
        writer.Double(imu_info_.yaw);
        writer.Key("pitch");
        writer.Double(imu_info_.pitch);
        writer.Key("roll");
        writer.Double(imu_info_.roll);
    }
    writer.EndObject();
    std::string result = buf_json.GetString();
    const char* topic="MACH/RCOut";
    if (!connected_.load()) {
        return;
    }
    publish(nullptr,topic,result.size(),result.data());
}

void Mqtt_imp::SetBatteryStatus(const BatteryStatus& battery_status) {
    rapidjson::StringBuffer buf_json;
    rapidjson::PrettyWriter<rapidjson::StringBuffer> writer(buf_json);
    {
        battery_status_ = battery_status;
        writer.StartObject();
        writer.Key("voltage");
        writer.Double(battery_status_.voltage);
        writer.Key("current");
        writer.Double(battery_status_.current);
        writer.Key("remaining");
        writer.Double(battery_status_.remaining);
    }
    writer.EndObject();
    std::string result = buf_json.GetString();
    const char* topic = "MACH/BatteryStatus";
    publish(nullptr, topic, result.size(), result.data());
}

void Mqtt_imp::SetAutoCruisePtr(std::shared_ptr<rov_planning::AutoCruise>& ptr)
{
	auto_cruise_ = ptr;
}

}  // namespace MQTT
